sdk/dotnet/Pulumi.Automation/Runtime/LanguageRuntimeService.cs

```diff
@@ -23,10 +24,69 @@ public LanguageRuntimeService(CallerContext callerContext)
             this._callerContext = callerContext;
         }
 
-        public override Task<GetRequiredPluginsResponse> GetRequiredPlugins(GetRequiredPluginsRequest request, ServerCallContext context)
+        public override async Task<GetRequiredPluginsResponse> GetRequiredPlugins(GetRequiredPluginsRequest request, ServerCallContext context)
         {
             var response = new GetRequiredPluginsResponse();
-            return Task.FromResult(response);
+            var assembly = this._callerContext.Program.DiscoveryAssembly;
+
+            // if discovery assembly is null, assuming consumer installed plugins manually
+            if (assembly is null)
+                return response;
+
+            var pulumiAssemblyNames = assembly.GetReferencedAssemblies().Where(x => x.Name != null && x.Name.StartsWith("Pulumi.")).ToArray();
+            foreach (var assemblyName in pulumiAssemblyNames)
+            {
+                var pulumiAssembly = Assembly.Load(assemblyName);
+
+                // get version.txt
+                // this part would probably be less error prone if there was an assembly attribute available
+                // instead of using the embedded version.txt resource
+                // pulumiAssembly.GetCustomAttribute<ResourcePluginAttribute>();
```

Comment:
Looking for your thoughts on this, @mikhailshilkov 

I think an assembly attribute would be a lot less error prone than trying to retrieve the embedded `version.txt` resource. We could put it in the `.csproj` for the resource provider packages like:
```
<ItemGroup>
  <AssemblyAttribute Include="Pulumi.ResourcePluginAttribute">
    <_Parameter1>$(Version)</_Parameter1>
  </AssemblyAttribute>
</ItemGroup>
```

And then the version could be inserted on build the same way that you are creating the `version.txt` file on build, I believe. Would probably still keep `version.txt` around since I think the GO dotnet language host would still need it, but just a thought. I know it may be a big lift to add that to everything.